src/controller/CHIPDeviceControllerFactory.h

```diff
@@ -209,6 +209,9 @@ class DeviceControllerFactory
     //
     bool ReleaseSystemState();
 
+    // Like RetainSystemState(), but will re-initialize the system state first if necessary.
```

Comment:
It's "necessary" if it's been shut down, right?

Should this be called something like EnsureRetainedSystemState()?  Feels like the important part here is that once this function returns success we have a "live" system state that has been retained, right?

And separately, should we have a `EnsureSystemState()` which does the "init if necessary" bits and then consumers just call retain separately?  It's not 100% clear why we are rolling both operations into one function.  Maybe more comments explaining how this should be used would help?

Reply:
I'm mainly adding this because I didn't want to change the behavior of the existing `RetainSystemState` to potentially re-initialize. `EnsureRetainedSystemState` is maybe a bit ambiguous... if the retain count is already > 0 then the system state is already retained and it sounds like it would then do nothing?

We have the equivalent of `EnsureSystemState` internally, but I don't think we want to publicly expose the ability to do that without also taking a retain count at the same time -- after all the retain count is what is meant to manage the lifecycle of the system state, and the `+0` state of being initialized but not retained yet is something that only exists transiently while bringing up a controller.

Reply:
That makes sense.